File: java/src/main/java/adventofcode/day2/PasswordPhilosophy.java
package adventofcode.day2;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

// Day 2 of Advent of Code challenge
// Link: https://adventofcode.com/2020/day/2
public class PasswordPhilosophy {

  public static int solution(int part) {
    List<String> lines = parseInput();
    int result = 0;

    for (String line : lines) {
      PasswordPolicy passwordPolicy = getPasswordPolicy(line);
      boolean isValidPassword = part == 1
              ? isValidPasswordPart1(passwordPolicy) : isValidPasswordPart2(passwordPolicy);
      result += isValidPassword ? 1 : 0;
    }

    return result;
  }

  private static boolean isValidPasswordPart1(PasswordPolicy passwordPolicy) {
    int count = 0;

    for (char c : passwordPolicy.password.toCharArray()) {
      if (c == passwordPolicy.value) count++;
    }

    return count <= passwordPolicy.maximum && count >= passwordPolicy.minimum;
  }

  private static boolean isValidPasswordPart2(PasswordPolicy passwordPolicy) {
    boolean result = false;
    if (passwordPolicy.password.charAt(passwordPolicy.minimum - 1) == passwordPolicy.value) result = true;
    if (passwordPolicy.password.charAt(passwordPolicy.maximum - 1) == passwordPolicy.value) result = !result;
    return result;
  }

  private static PasswordPolicy getPasswordPolicy(String input) {
    String[] segments = input.split(":");
    String[] passwordPolicySegment = segments[0].split(" ");
    String[] minMax = passwordPolicySegment[0].split("-");

    return new PasswordPolicy(passwordPolicySegment[1].charAt(0),
            Integer.parseInt(minMax[0]),Integer.parseInt(minMax[1]), segments[1].trim());
  }

  private static List<String> parseInput() {
    List<String> result = new ArrayList<>();

    try {
      FileReader fileReader = new FileReader("java/src/main/resources/inputs/day2.txt");
      BufferedReader bufferedReader = new BufferedReader(fileReader);
      String line = bufferedReader.readLine();
      while (line != null) {
        result.add(line);
        line = bufferedReader.readLine();
      }
    } catch (IOException e) {
      e.printStackTrace();
    }

    return result;
  }

  public static void main(String[] args) {
    System.out.println("Part 1: Number of valid passwords: " + solution(1));
    System.out.println("Part 2: Number of valid passwords: " + solution(2));
  }
}

File: golang/day2/password_philosophy.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

// Day 2 of Advent of Code challenge
// Link: https://adventofcode.com/2020/day/2

// PasswordPolicy to represent password policies
type PasswordPolicy struct {
	value    byte
	minimum  int64
	maximum  int64
	password string
}

func parseInput() []string {
	var result []string
	inputFile, error := os.Open("input.txt")
	if error != nil {
		log.Fatal(error)
	}
	defer inputFile.Close()

	scanner := bufio.NewScanner(inputFile)
	for scanner.Scan() {
		result = append(result, scanner.Text())
	}

	if error := scanner.Err(); error != nil {
		log.Fatal(error)
	}

	return result
}

func getPassWordPolicy(input string) *PasswordPolicy {
	var segments []string = strings.Split(input, ":")
	var passwordPolicySegment []string = strings.Split(segments[0], " ")
	var minMax []string = strings.Split(passwordPolicySegment[0], "-")
	min, _ := strconv.ParseInt(minMax[0], 10, 64)
	max, _ := strconv.ParseInt(minMax[1], 10, 64)

	return &PasswordPolicy{passwordPolicySegment[1][0], min, max, strings.Trim(segments[1], " ")}
}

func solution(part int) int {
	var lines []string = parseInput()
	var result int = 0

	for _, line := range lines {
		passwordPolicy := getPassWordPolicy(line)
		var isValid bool = false
		if part == 1 {
			isValid = isValidPasswordPart1(passwordPolicy)
		} else {
			isValid = isValidPasswordPart2(passwordPolicy)
		}

		if isValid {
			result++
		}
	}

	return result
}

func isValidPasswordPart1(passwordPolicy *PasswordPolicy) bool {
	var count int64 = 0
	for _, char := range passwordPolicy.password {
		if byte(char) == passwordPolicy.value {
			count++
		}
	}
	return count >= passwordPolicy.minimum && count <= passwordPolicy.maximum
}

func isValidPasswordPart2(passwordPolicy *PasswordPolicy) bool {
	var result bool = false
	if passwordPolicy.password[passwordPolicy.minimum-1] == passwordPolicy.value {
		result = true
	}
	if passwordPolicy.password[passwordPolicy.maximum-1] == passwordPolicy.value {
		result = !result
	}

	return result
}

func main() {
	var result1 = solution(1)
	var result2 = solution(2)
	fmt.Printf("Part 1: Number of valid passwords: %d", result1)
	fmt.Println("")
	fmt.Printf("Part 2: Number of valid passwords: %d", result2)
}
